oci8.go

```diff
@@ -1636,15 +1654,30 @@ func ociGetErrorS(err unsafe.Pointer) error {
 	return errors.New(s)
 }
 
+// isBadConnection checks the error string for ORA errors that would mean the connection is bad
 func isBadConnection(error string) bool {
-	if len(error) <= 8 {
+	if len(error) < 9 || error[0:4] != "ORA-" {
+		// if error is less than 9 and is not an ORA error
 		return false
 	}
-	errorCode := error[0:9]
-	for _, badConnCode := range badConnCodes {
-		if badConnCode == errorCode {
-			return true
-		}
+	// only check number part, ORA is already checked
+	switch error[4:9] {
+	/*
+		bad connection errors:
+		ORA-00028: your session has been killed
+		ORA-01012: Not logged on
+		ORA-01033: ORACLE initialization or shutdown in progress
+		ORA-01034: ORACLE not available
+		ORA-01089: immediate shutdown in progress - no operations are permitted
+		ORA-03113: end-of-file on communication channel
+		ORA-03114: Not Connected to Oracle
+		ORA-03135: connection lost contact
+		ORA-12528: TNS:listener: all appropriate instances are blocking new connections
+		ORA-12537: TNS:connection closed
+	*/
+	case "00028", "01012", "01033", "01034", "01089", "03113", "03114", "03135", "12528", "12537":
```

Comment:
ODPI-C (which is used in some other Oracle DB APIs like [cx_Oracle](https://github.com/oracle/python-cx_Oracle), [node-oracledb](https://github.com/oracle/node-oracledb/) and [go-goracle](https://github.com/go-goracle/goracle/)) checks for these connection error codes : https://github.com/oracle/odpi/blob/v2.4.2/src/dpiError.c#L69  
These explicit numbers supplement the internal check https://github.com/oracle/odpi/blob/v2.4.2/src/dpiConn.c#L202-L208